### PR TITLE
Add timeout to Finality API #380

### PR DESCRIPTION
**Pull request by adecaro**

### CODE REVIEW

**Comment by adecaro** [comment]
Linked to https://github.com/hyperledger-labs/fabric-token-sdk/pull/384

File: platform/fabric/core/generic/chaincode/support.go [hunk 1]
```diff
@@ -51,7 +53,7 @@ type Channel interface {
 	NewPeerClientForIdentity(peer view.Identity) (peer.Client, error)
 
 	// IsFinal takes in input a transaction id and waits for its confirmation.
```
Comment (HagarMeir) [inline platform/fabric/core/generic/chaincode/support.go:55]: Please update comment

File: platform/fabric/core/generic/committer/committer.go [hunk 2]
```diff
@@ -132,12 +130,12 @@ func (c *committer) Commit(block *common.Block) error {
 }
 
 // IsFinal takes in input a transaction id and waits for its confirmation.
-func (c *committer) IsFinal(txid string) error {
-	c.metrics.EmitKey(0, "committer", "start", "IsFinal", txid)
-	defer c.metrics.EmitKey(0, "committer", "end", "IsFinal", txid)
+func (c *committer) IsFinal(ctx context.Context, txID string) error {
```
Comment (HagarMeir) [inline platform/fabric/core/generic/committer/committer.go:133]: Please update comment

File: platform/fabric/core/generic/committer/committer.go [hunk 3]
```diff
@@ -308,7 +306,10 @@ func (c *committer) listenTo(txid string, timeout time.Duration) error {
 	for i := 0; i < iterations; i++ {
 		timeout := time.NewTimer(c.pollingTimeout)
 
+		stop := false
 		select {
+		case <-ctx.Done():
+			stop = true
```
Comment (HagarMeir) [inline platform/fabric/core/generic/committer/committer.go:312]: should there also be a `timeout.Stop()` here?

File: platform/fabric/core/views/finality.go [hunk 4]
```diff
@@ -52,7 +54,7 @@ func (i *IsFinalResponderView) Call(context view.Context) (interface{}, error) {
 	var ch driver.Channel
 	ch, err = network.Channel(isFinalRequest.Channel)
 	if err == nil {
-		err = ch.IsFinal(isFinalRequest.TxID)
+		err = ch.IsFinal(context.Background(), isFinalRequest.TxID)
```
Comment (HagarMeir) [inline platform/fabric/core/views/finality.go:57]: why not pass `ctx` here?

File: platform/fabric/driver/finality.go [hunk 5]
```diff
@@ -6,11 +6,15 @@ SPDX-License-Identifier: Apache-2.0
...
 
-import "github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+import (
+	"context"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+)
 
 type Finality interface {
 	// IsFinal takes in input a transaction id and waits for its confirmation.
```
Comment (HagarMeir) [inline platform/fabric/driver/finality.go:16]: Update comment

File: platform/orion/core/generic/committer/committer.go [hunk 6]
```diff
@@ -349,7 +350,10 @@ func (c *committer) listenToFinality(txID string, timeout time.Duration) error {
 	for i := 0; i < iterations; i++ {
 		timeout := time.NewTimer(c.pollingTimeout)
 
+		stop := false
 		select {
+		case <-ctx.Done():
+			stop = true
```
Comment (HagarMeir) [inline platform/orion/core/generic/committer/committer.go:356]: also here `timeout.Stop()` is not missing?

File: platform/orion/driver/finality.go [hunk 7]
```diff
@@ -6,11 +6,15 @@ SPDX-License-Identifier: Apache-2.0
...
 
-import "github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+import (
+	"context"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+)
 
 type Finality interface {
 	// IsFinal takes in input a transaction id and waits for its confirmation.
```
Comment (HagarMeir) [inline platform/orion/driver/finality.go:16]: update comment

File: platform/fabric/core/generic/committer/committer.go (hunk 3 above)
Comment (adecaro) [reply platform/fabric/core/generic/committer/committer.go:312]: good catch

File: platform/fabric/core/views/finality.go (hunk 4 above)
Comment (adecaro) [reply platform/fabric/core/views/finality.go:57]: I think the code is outdated.

**Comment by adecaro** [comment]
@HagarMeir, thanks for the catches. I have updated the PR as suggested :)

**Review by HagarMeir — APPROVED** [review]

**Pull request merged**